apps/src/javalab/Backpack.jsx

```diff
@@ -5,29 +5,59 @@ import color from '@cdo/apps/util/color';
 import onClickOutside from 'react-onclickoutside';
 import JavalabButton from './JavalabButton';
 import javalabMsg from '@cdo/javalab/locale';
-
-const placeholderFiles = [
-  'MyClass.java',
-  'MyPainter.java',
-  'NeighborhoodMain.java'
-];
+import {connect} from 'react-redux';
+import {setSource} from './javalabRedux';
 
 /**
  * A button that drops down to a set of importable files, and closes itself if
  * you click on the import button, or outside of the dropdown.
  */
-export class Backpack extends Component {
+class Backpack extends Component {
   static propTypes = {
     isDarkMode: PropTypes.bool.isRequired,
-    isDisabled: PropTypes.bool.isRequired
+    isDisabled: PropTypes.bool.isRequired,
+    onImport: PropTypes.func.isRequired,
+    // populated by redux
+    backpackApi: PropTypes.object.isRequired
   };
 
   state = {
-    dropdownOpen: false
+    dropdownOpen: false,
+    backpackFilenames: [],
+    backpackFilesLoading: false,
+    backpackLoadError: false,
+    selectedFiles: []
   };
 
   expandDropdown = () => {
-    this.setState({dropdownOpen: true});
+    this.setState({
+      dropdownOpen: true,
+      backpackLoadError: false,
+      selectedFiles: [],
+      backpackFilenames: []
+    });
+    if (this.props.backpackApi.hasBackpack()) {
+      this.setState({backpackFilesLoading: true});
+      this.props.backpackApi.getFileList(
+        this.onFileListLoadError,
+        this.onFileListLoadSuccess
+      );
+    }
```

Comment:
What do users see if the backpack API isn't available? would it just be empty?

Reply:
if the api has a load error they see the error message. If they haven't added anything to the backpack yet/we got a null channel id from the backend they see you have no files.